### ISSUE TITLE
Request for adding more benchmark & question about optimized-rnnt-loss

### ISSUE THREAD

**Issue by maxwellzh** [csukuangfj/transducer-loss-benchmarking#10]
Hi, Fangjun @csukuangfj , can you add the benchmark on https://github.com/1ytic/warp-rnnt ? `warp-rnnt` says it's faster than `warp-transducer` (non-optimized ver.)

And I have a question about the benchmarking results. It's shown that whatever w/ or w/o sorted utterances, although `optimized_transducer` consumes less memory than default `warp-transducer`, the step time is more. Why is that? I think the tricks in [Improving RNN Transducer Modeling for End-to-End Speech Recognition](https://arxiv.org/abs/1909.12415) should theoratically fasten the training.

Thanks.

Name | Average step time (us) | Peak memory usage (MB)
-- | -- | --
optimized_transducer | 376954 | 7495.9
warp-transducer | 275852 | 19072.6

Name | Average step time (us) | Peak memory usage (MB)
-- | -- | --
optimized_transducer | 567684 | 10903.1
warp-transducer | 210772 | 13061.8

**Comment by csukuangfj** [comment]
I just added the benchmark results for warp-rnnt.

> the step time is more. Why is that?

To explain that, I think it needs a more detailed analysis with some profiling tools, e..g, `nvprof`.
I suspect the potential reason is that it needs extra data structure and memory access to get the information to answer:
- which batch does frame 'i' come from?

Since paddings are removed and we pass a 2-D tensor to compute the loss, you have to use extra data to recover the batch information from the frame index.

**Comment by maxwellzh** [comment]
Thanks for the clarification. So the performance (speed) degradation is possibly due to the implementation of accessing data in the squeezed layout.

And btw, since the most common joint network in RNN-T consist of FC and activation layers only, I think it would bring even more efficiency gain if we squeeze the 4-dim tensor before the FC layer.

**Comment by csukuangfj** [comment]
> And btw, since the most common joint network in RNN-T consist of FC and activation layers only, I think it would bring even more efficiency gain if we squeeze the 4-dim tensor before the FC layer.

That is how the benchmark code is doing at present. Please see
https://github.com/csukuangfj/transducer-loss-benchmarking/blob/8373897fd86a64425baa198247cde439f10b1bd2/benchmark_ot.py#L122-L140

Also, you can find its usage in ASR training in
https://github.com/k2-fsa/icefall/blob/master/egs/librispeech/ASR/transducer_stateless/joiner.py#L29

**Comment by maxwellzh** [comment]
Thanks for the quick reply. Seems I just miss this.

I've impl a cuda version of the squeezing and the broadcasting sum https://github.com/maxwellzh/torch-gather It should be faster than the seq-by-seq `torch.cat` (Edit: may only compatible with torch < 1.11, since 1.11 has some c++ header deprecated)
